### DISCUSSION TITLE
State of the OpenDataCam Project - Q3 2022

### DISCUSSION THREAD

**Discussion by vsaw** [opendatacam/opendatacam#565] [Announcements]
Hi everyone,

I'm Valentin, the maintainer of OpenDataCam. In this post I would like to give you an update on the current state of OpenDataCam and why things have been moving slowly in the last few years.

I'll try to give you a broad overview over different areas of the Project including

- Maintainers & Developers
- Community
- Technology & Roadmap
- Outlook

If there is anything else you would like to discuss or have questions, please leave a comment.

## Maintainers & Developers

At the moment I am the only "active" maintainer and developer of OpenDataCam. Like many open source maintainers, I'm working on OpenDataCam in my spare time for free. Which is why my work on OpenDataCam is going slow & very irregular. I would love to dedicate more of my time for OpenDataCam, but for this I would need OpenDataCam to generate a bit of stable income for me. To achieve this, I am trying to offer [paid professional services for OpenDataCam](https://nono.cam/opendatacam/), like pre-installed Hardware & custom software development (the majority of OpenDataCam remains FOSS). While I managed to do a few projects this way, it is unfortunately still not enough for me to dedicate more time to OpenDataCam.

## Community

I'm happy to say that despite all the challenges, the OpenDataCam community continues to grow!

![image](https://user-images.githubusercontent.com/829308/196673532-46a7b684-e269-4ce3-a6f0-ec23646d8fbf.png)

The amount of GitHub Stars has grown by ~100 in this year alone. Over the last 14 days. OpenDataCam has seen 2,005 Views and 419 Unique visitors on GitHub. And there is a small but steady stream of activity in the OpenDataCam discussions here on GitHub.

In addition to GitHub, OpenDataCam has a Twitter Handle as well [@opendatacam](https://twitter.com/opendatacam) but for the reasons mentioned above, I don't find much time to post on Twitter, hence the very few followers.

Outside of GitHub and Twitter, I keep advocating for OpenDataCam, for example by giving a talk about OpenDataCam on the Internet of Things Conference in Munich earlier this year ([Slides](https://drive.google.com/file/d/1OMmx__eLrR2q0DYaXc9GZiZgY7QIsRoC/view)). But as with everything else, time is the limiting factor.

## Technology & Roadmap

From a technology perspective, OpenDataCam is falling behind on the State of the Art. The use of [Darknet](https://github.com/opendatacam/darknet) and Yolo v4 is keeping performance and detection results behind newer implementations such as [YOLOv7](https://github.com/WongKinYiu/yolov7). On the Tracking side, [node-moving-things-tracker](https://github.com/opendatacam/node-moving-things-tracker) has severe issues tracking objects in busy scenes and can not resume tracks if objects become hidden (e.g. car driving behind a tree). All of these issues are solved in State of the Art trackers like [DeepSort](https://github.com/ZQPei/deep_sort_pytorch). Upgrading OpenDataCam to take advantages of newer models and algorithms is difficult, for three reasons

1. OpenDataCam uses Darknet, most of the State of the Art AI research however happens in PyTorch. So we either need to port it to Darknet ourselves or hope someone does the porting (both unlikely)
1. OpenDataCam code is deeply tied to Darknet. Switching to PyTorch or similar to get easy access to newer models would require significant refactoring.
1. Those libraries are almost exclusively written in Python, while OpenDataCam is written in JavaScript. Therefore we can not use the code directly but would need re-write OpenDataCam or the Libraries.

In summary OpenDataCam is falling behind and possibly even reached a point where further improvements are difficult to do without significant development effort (or a rewrite).

Which is why I'm a bit hesitant to put much time into the OpenDataCam v4 release right now. Yes there are improvements in the development branch like the "Area Counter", but if there is not a clear path forward to keep up with the other development in the computer vision field, I wonder if OpenDataCam simply will be "not good enough" in the near future. That is why I do not communicate a release date for OpenDataCam v4 at the moment.

## Outlook

What's next for OpenDataCam? My primary focus right now is to see if I can generate enough revenue with OpenDataCam so I can become an OpenDataCam maintainer full time. For this I spend most of my time at the moment talking to different members of the community and potential clients in understanding how I could create a paid offering that fits the OpenDataCam project.

I will keep being available for the OpenDataCam community in GitHub and answering as many questions as I can, but there is likely going to be little development from me in the future.

To improve this situation I'm hoping for help of the OpenDataCam community. Here are a few ideas what you can do:

- **Spread the word!** Recommend [OpenDataCam Professional Services](https://nono.cam/opendatacam/) to businesses and other people that are looking for a commercial solution.
- **Share your thoughts!** If you have an idea how OpenDataCam could generate revenue so that people like me can work on it full time, please reach out to me! (Twitter: https://twitter.com/_vsaw, Email: valentin@nono.cam)
- **Become a contributor!** There are many ways to contribute to OpenDataCam (Showcasing what you do with OpenDataCam, improving Documentation like setup instructions, or writing code) and all of them are welcome!

Thanks for making it this far and let's keep counting! ;-)

**Comment by tdurand** [comment]
Hey @vsaw , 

Thanks for all the energy you are putting in the project ! 

On my side, I would probably be interested if some funding is secured to do some development work and improvement on ODC, but right now I can't afford to dedicate to it on my free time, nor I have the bandwidth to look for funding etc etc...

On the technical side, yes it is falling a bit behind state of the art, switching to the latest Yolov7 + tracker stuff would be needed and require significant effort as we would need to switch to python backend.. , but I guess it's doable and not too bad as tooling has matured..

Best

Thibault

**Comment by vsaw** [comment]
New version posted here https://github.com/opendatacam/opendatacam/discussions/603
